library/Haste/Ajax/ReloadHelper.php

```diff
@@ -0,0 +1,263 @@
+     */
+    public static function updateContentElementBuffer(ContentModel $model, $buffer)
+    {
+        if ($model->type === 'module' && ($module = ModuleModel::findByPk($model->module)) !== null) {
+            return static::updateFrontendModuleBuffer($module, $buffer);
+        }
+
+        $events = [];
+
+        foreach (static::$elements as $event => $elements) {
+            if (in_array((int)$model->id, $elements, true)) {
+                $events[] = $event;
+            }
+        }
+
+        if (count($events) > 0) {
+            $buffer = static::addDataAttributes($buffer, static::getModelKey($model), $events);
+        }
+
+        return $buffer;
+    }
+
+    /**
+     * Subscribe the frontend module
+     *
+     * @param int   $id
+     * @param array $events
+     */
+    public static function subscribeFrontendModule($id, array $events)
+    {
+        $id = (int)$id;
+
+        foreach ($events as $event) {
+            if (!static::$modules[$event] || !in_array($id, static::$modules[$event], true)) {
+                static::$modules[$event][] = $id;
+            }
+        }
+    }
+
+    /**
+     * Store the frontend module in the response if applicable
+     *
+     * @param array       $events
+     * @param ModuleModel $module
+     * @param string      $buffer
+     */
+    public static function storeFrontendModuleResponse(array $events, ModuleModel $module, $buffer)
+    {
+        static::storeResponse($events, static::$modules, $module, $buffer);
+    }
+
+    /**
+     * Update the frontend module buffer
+     *
+     * @param ModuleModel $model
+     * @param string      $buffer
+     *
+     * @return string
+     */
+    public static function updateFrontendModuleBuffer(ModuleModel $model, $buffer)
+    {
+        $events = [];
+
+        foreach (static::$modules as $event => $modules) {
+            if (in_array((int)$model->id, $modules, true)) {
+                $events[] = $event;
+            }
+        }
+
+        if (count($events) > 0) {
+            $buffer = static::addDataAttributes($buffer, static::getModelKey($model), $events);
+        }
+
+        return $buffer;
+    }
+
+    /**
+     * Return true if there are listeners
+     *
+     * @return bool
+     */
+    public static function hasListeners()
+    {
+        return count(static::$modules) > 0 || count(static::$elements) > 0;
+    }
+
+    /**
+     * Get the response
+     *
+     * @return JsonResponse|null
+     */
+    public static function getResponse()
+    {
+        if (count(static::$response) < 1) {
+            return null;
+        }
+
+        return new JsonResponse(array_values(static::$response));
+    }
+
+    /**
+     * Add the data attributes
+     *
+     * @param string $buffer
+     * @param string $id
+     * @param array  $events
+     *
+     * @return string
+     */
+    private static function addDataAttributes($buffer, $id, array $events)
+    {
+        $dom = new \DOMDocument();
```

Comment:
So normally I'd say it is a great choice to use `DOMDocument`. However, it's just a fact that I have experienced too many issues when using this. Especially when I was using it for my css-class-replacer extension of which David eventually extracted this: https://github.com/netzmacht/php-dom-manipulator/blob/master/src/Netzmacht/DomManipulator/Converter/DomDocumentConverter.php
You can see that there's a whole lot of stuff done to prevent encoding errors etc.
I was just wondering if we cannot use a regex in this case. I mean the place we're looking for is the first space after an opening `<` if there is any so we can insert our data attributes. I think this should work.